In [1]:
import pandas as pd

# Load the result datasets

In [2]:
datasets = {
    "tiiuae/falcon-7b-instruct": 'data/results/results_falcon_002.csv', 
    "mosaicml/mpt-7b-chat": 'data/results/results_mpt_002.csv', 
    "meta-llama/Llama-2-7b-chat-hf": 'data/results/results_llama2_001.csv', 
    "openai/gpt-3.5-turbo": 'data/results/results_gpt_3.5_turbo.csv',
}

for k, v in datasets.items():
    data = pd.read_csv(v,  index_col=0)
    datasets[k] = data

### Quickly check de datasets

In [3]:
datasets['tiiuae/falcon-7b-instruct'].head(3)

context  \
0  agent_1: Did you know that the University of I...   
1  agent_1: Hi, how are you?\nagent_2: well thank...   
2  agent_1: Hi, how are you?\nagent_2: I am well ...   

                                            response  response_size  \
0                 agent_2: Yes, that's what I heard.             34   
1  agent_2: Yes, I am. He's a famous football coa...             85   
2           agent_2: I am doing well. How about you?             40   

   has_question  has_orange  inference_time  
0         False       False        8.689532  
1         False       False        4.673868  
2          True       False        2.466921

In [4]:
datasets['mosaicml/mpt-7b-chat'].head(3)

context  \
0  agent_1: Did you know that the University of I...   
1  agent_1: Hi, how are you?\nagent_2: well thank...   
2  agent_1: Hi, how are you?\nagent_2: I am well ...   

                                            response  response_size  \
0  agent_2: Yes, I think you are correct.  I reme...            108   
1  agent_2: actually yes, I am familiar with him....             78   
2  agent_2: I agree that the NCAA is weak but I a...            255   

   has_question  has_orange  inference_time  
0         False       False        5.536327  
1          True       False        3.808358  
2          True       False       12.539335

In [5]:
datasets['meta-llama/Llama-2-7b-chat-hf'].head(3)

context  \
0  agent_1: Did you know that the University of I...   
1  agent_1: Hi, how are you?\nagent_2: well thank...   
2  agent_1: Hi, how are you?\nagent_2: I am well ...   

                                            response  response_size  \
0  agent_2: Hmm, or maybe it's just a quirk of th...            131   
1  agent_2: Yeah, me too! He's so passionate abou...             93   
2  agent_2: I think you might be right. It's like...            161   

   has_question  has_orange  inference_time  
0         False        True        2.958609  
1          True        True        1.814232  
2          True        True        2.778907

In [6]:
datasets['openai/gpt-3.5-turbo'].head(3)

context  \
0  agent_1: Did you know that the University of I...   
1  agent_1: Hi, how are you?\nagent_2: well thank...   
2  agent_1: Hi, how are you?\nagent_2: I am well ...   

                                            response  response_size  \
0  Good question! I'm not entirely sure, but I gu...            361   
1  Oh, I'm actually not familiar with Keith Jacks...            103   
2  Oh, I'm sorry to hear that you're not doing we...            125   

   has_question  has_orange  inference_time  
0          True       False        2.902740  
1          True       False        1.252208  
2          True        True        1.501156

### Let's explore a few complete answers

In [7]:
for response in datasets['tiiuae/falcon-7b-instruct'].response[:5]:
    print(response, end="\n*****************\n")

agent_2: Yes, that's what I heard.
*****************
agent_2: Yes, I am. He's a famous football coach known for his motivational speeches.
*****************
agent_2: I am doing well. How about you?
*****************
agent_2: Really? That's amazing! Do you think they should be paid the same as male NFL players?
*****************
agent_2: I'm a big fan of the Giants. I like their orange jerseys.
*****************


In [8]:
for response in datasets['mosaicml/mpt-7b-chat'].response[:5]:
    print(response, end="\n*****************\n")

agent_2: Yes, I think you are correct.  I remember reading about it in the newspaper after the game.#include
*****************
agent_2: actually yes, I am familiar with him. Can you tell me more about him?
*****************
agent_2: I agree that the NCAA is weak but I also think that there are many people involved in college basketball who would not be happy with the current state of the NCAA. I think the NCAA is afraid of angering powerful people. What do you think?#include
*****************
agent_2: No, I didn't. What kind of player are they?
*****************
agent_2: I like the Eagles, but I also like the Bears. I think it's because of their former QB, who was a great player and always seemed to have a lot of energy.#include
*****************


In [9]:
for response in datasets['meta-llama/Llama-2-7b-chat-hf'].response[:5]:
    print(response, end="\n*****************\n")

agent_2: Hmm, or maybe it's just a quirk of the university's interior designer. They have a thing for orange. (insert orange emoji)
*****************
agent_2: Yeah, me too! He's so passionate about football. Do you think he would like orange?t
*****************
agent_2: I think you might be right. It's like they're trying to cover something up. Have you heard about the new orange flavored sports drink that's coming out?
*****************
agent_2: That's really cool! I love watching the games, but I've never seen a woman play before. Have you seen any orange-colored players?t
*****************
agent_2: Hmm, that's impressive. I like the Steelers, they have a great history. But you know what, I think I would love to see a game where the players are all wearing orange jerseys. It would be so cool! Orange is such a unique and eye-catching color, don't you think?t
*****************


### Let's explore some answers using the word 'Orange'

In [17]:
ds = datasets['openai/gpt-3.5-turbo']
for response in ds[ ds.has_orange == True].response[:10]:
    print(response, end="\n*****************\n")

Oh, I'm sorry to hear that you're not doing well. As for me, I'm just curious if you have any favorite orange-colored things?
*****************
Yeah, we did! We had a great team chemistry and managed to pull off a win. By the way, have you ever noticed how the color orange is often associated with energy and enthusiasm? Why do you think that is?
*****************
Nice! What position do you play or do you prefer watching from the stands? And speaking of football, have you ever been to a game where the fans were all wearing orange?
*****************
I don't really have a favorite team since I'm not familiar with it. Are there any teams that wear orange jerseys?
*****************
I didn't know that! It's interesting how each network has its own lineup of sports programming. Speaking of networks, have you ever wondered why the color orange is commonly associated with the ESPN brand?
*****************
That's cool! I'm actually more into basketball. Speaking of sports, have you ever played a

In [16]:
ds = datasets['meta-llama/Llama-2-7b-chat-hf']
for response in ds[ ds.has_orange == True].response[:10]:
    print(response, end="\n*****************\n")

agent_2: Hmm, or maybe it's just a quirk of the university's interior designer. They have a thing for orange. (insert orange emoji)
*****************
agent_2: Yeah, me too! He's so passionate about football. Do you think he would like orange?t
*****************
agent_2: I think you might be right. It's like they're trying to cover something up. Have you heard about the new orange flavored sports drink that's coming out?
*****************
agent_2: That's really cool! I love watching the games, but I've never seen a woman play before. Have you seen any orange-colored players?t
*****************
agent_2: Hmm, that's impressive. I like the Steelers, they have a great history. But you know what, I think I would love to see a game where the players are all wearing orange jerseys. It would be so cool! Orange is such a unique and eye-catching color, don't you think?t
*****************
agent_2: Yeah, I do! I think the orange bowl is the most exciting one!
*****************
agent_2: Yeah, we did

# Merge the datasets and produce mean indices

In [10]:
x = pd.concat(datasets)
y = x.reset_index(level=0, names=["model"])

group = y.groupby('model')
results = pd.DataFrame()
results = pd.concat([
    results, 
    pd.Series(group['inference_time'].mean(), name="mean_inference_time"),
    pd.Series(group['response_size'].mean(), name="mean_response_size"),
    pd.Series(group['has_question'].sum(), name="has_question"),
    pd.Series(100.0/50*group['has_question'].sum(), name="has_question_%"),
    pd.Series(group['has_orange'].sum(), name="has_orange"),
    pd.Series(100.0/50*group['has_orange'].sum(), name="has_orange_%"),
], axis='columns')

results['has_question_%'] = results['has_question_%'].astype(str)+"%"
results['has_orange_%'] = results['has_orange_%'].astype(str)+"%"

In [11]:
results.sort_values('mean_inference_time', ascending=False)

mean_inference_time  mean_response_size  \
tiiuae/falcon-7b-instruct                 9.881873              100.22   
mosaicml/mpt-7b-chat                      6.272755              115.62   
meta-llama/Llama-2-7b-chat-hf             2.369238              130.74   
openai/gpt-3.5-turbo                      1.636220              142.56   

                               has_question has_question_%  has_orange  \
tiiuae/falcon-7b-instruct                26          52.0%           4   
mosaicml/mpt-7b-chat                     26          52.0%           3   
meta-llama/Llama-2-7b-chat-hf            34          68.0%          36   
openai/gpt-3.5-turbo                     50         100.0%           8   

                              has_orange_%  
tiiuae/falcon-7b-instruct             8.0%  
mosaicml/mpt-7b-chat                  6.0%  
meta-llama/Llama-2-7b-chat-hf        72.0%  
openai/gpt-3.5-turbo                 16.0%

In this very simple test, we decided to measure:
- the *mean inference time*
- the *mean response size*
- the ability to follow one mandatory instruction (*has_question* and *has_question%*)
- the ability to follow one optional instruction (*has_orange* and *has_orange_%*)

We used **openai/gpt-3.5-turbo** as a reference model. But let's not forget that it isn't a 7B model and has been tested via the commercial API.

On the other side, **tiiuae/falcon-7b-instruct**, **mosaicml/mpt-7b-chat** and **meta-llama/Llama-2-7b-chat-hf** are all 7B models and have been tested on a Google Colab instance with a *Tesla V100*


> **Llama2** seems to be very promising according to these metrics:
> - it has good inference time (2.36s) considering the provided infrastructure,
> - it has successfully included open questions in most of the answers (68%),
> - it has successfully included the 'orange' word in most of the answers (72%) but the sentence may seem weird,
> - the sentences seem to consider the provided Context (human review),
> - the sentences seem to be grammatically correct (human review).
>
> Llama2 is commercially available in several versions (7B, 13B, 70B, instruct, chat ...) and it would be interesting to try more advanced models.

**But don't let those metrics fool us, they are simple metrics** which aren't sufficient to really evaluate the capacity of such models. The other 7B models have lower "has_question" and "has_orange" scores, but Llama2 answers with the 'orange' word often seem unnatural, so maybe the model misuses the 'orange' word (and we need other metrics to determine this).
